airbyte-cdk/python/airbyte_cdk/sources/streams/core.py

```diff
@@ -193,25 +193,54 @@ def read(  # type: ignore  # ignoring typing for ConnectorStateManager because o
 
                     checkpoint_interval = self.state_checkpoint_interval
                     checkpoint = checkpoint_reader.get_checkpoint()
-                    if checkpoint_interval and record_counter % checkpoint_interval == 0 and checkpoint is not None:
+                    if checkpoint_interval and record_counter % checkpoint_interval == 0 and checkpoint is not None and state_manager:
                         airbyte_state_message = self._checkpoint_state(checkpoint, state_manager=state_manager)
                         yield airbyte_state_message
 
                     if internal_config.is_limit_reached(record_counter):
                         break
             self._observe_state(checkpoint_reader)
             checkpoint_state = checkpoint_reader.get_checkpoint()
-            if checkpoint_state is not None:
+            if state_manager and checkpoint_state is not None:
                 airbyte_state_message = self._checkpoint_state(checkpoint_state, state_manager=state_manager)
                 yield airbyte_state_message
 
             next_slice = checkpoint_reader.next()
 
         checkpoint = checkpoint_reader.get_checkpoint()
-        if checkpoint is not None:
+        if state_manager and checkpoint is not None:
             airbyte_state_message = self._checkpoint_state(checkpoint, state_manager=state_manager)
             yield airbyte_state_message
 
+    def read_stateless(  # type: ignore  # ignoring typing for ConnectorStateManager because of circular dependencies
+        self,
+        connector_state_manager=None,
```

Comment:
This is something of a frustrating pattern we have. I can't actually instantiate a `ConnectorStateManager` instance within this method because I can't import it due to a circular dependency between `ConnectorStateManager` and the `Stream` class.

This is starting poke at us a bit and become an annoying pattern so we may want to see if we want to refactor or adjust the code in `connector_state_manager.py` to not reference the Stream class. But for now, not sure of a good way to avoid this parameter.

Reply:
can you move or copy this comment to the code so we have a trace. that's a pretty big gotcha